### PR TITLE
feat(ODC-63): add build tool and language detector for GitHub

### PR DESCRIPTION
**Pull request by MatousJobanek**
This PR is the first step for providing a support for build tool & language detection. Contains implementation supporting GitHub repositories reaching via GitHub API. As the authenticaiton options this PR introduces token and username+password based ways.

Support for the generic git and for other main git vendors such as GitLab and Bitbucket

### CODE REVIEW

File: pkg/git/repository/service.go [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+)
+
+type Service interface {
+	GetListOfFilesInRootDir() ([]string, error)
+	GetLanguageList() ([]string, error)
+}
+
+type ServiceCreator func(gitSource *git.Source) (Service, error)
+
+func NewService(gitSource *git.Source, serviceCreators []ServiceCreator) (Service, error) {
```
Comment (sbryzak) [inline pkg/git/repository/service.go]: This name is a little abstract, could we call it `NewGitService` or something more specific instead?

File: pkg/git/source.go [hunk 2]
```diff
@@ -0,0 +1,15 @@
+package git
```
Comment (alexeykazakov) [inline pkg/git/source.go]: Any particular reason not to use git source from the api package?

File: pkg/git/repository/service.go (hunk 1 above)
Comment (MatousJobanek) [reply pkg/git/repository/service.go]: Good point - looks better now?: 19eb36e

File: pkg/git/detector/detector.go [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+		}
+	}
+
+	return &BuildEnvStats{
+		SortedLanguages:    languageList,
+		DetectedBuildTools: environments,
+	}, nil
+}
+
+func detectBuildEnvironments(service repository.Service, detectedBuildTools chan *DetectedBuildTool) error {
```
[the remark this answers is not in the document]
Comment (MatousJobanek) [reply pkg/git/detector/detector.go]: good idea :-), thanks b8599c1

File: pkg/git/detector/detector.go [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+
+	"github.com/redhat-developer/git-service/pkg/git"
+	"github.com/redhat-developer/git-service/pkg/git/repository/github"
+)
+
+var gitServiceCreators = []repository.ServiceCreator{
+	github.NewRepoServiceIfMatches(),
+}
+
+func DetectBuildEnvironments(gitSource *git.Source) (*BuildEnvStats, error) {
```
Comment (sbose78) [inline pkg/git/detector/detector.go]: @MatousJobanek , you would be adding the 'output' of the detection to the `status` field of the `GitSource` CR, right?

**Comment by sbose78** [comment]
As a 'first' pass, we should be good if we handled Github , Bitbucket and Gitlab.

File: pkg/git/source.go (hunk 2 above)
Comment (MatousJobanek) [reply pkg/git/source.go]: The reason was that I didn't realize that the GitSource is already part of this repo :-). Thanks for pointing it out.
The actual integration with the controller (and reading of the secrets) will be part of different PR

File: pkg/git/detector/detector.go (hunk 4 above)
Comment (MatousJobanek) [reply pkg/git/detector/detector.go]: Yup, most likely there, yes. But not part of this PR :-)

**Comment by MatousJobanek** [comment]
> As a 'first' pass, we should be good if we handled Github , Bitbucket and Gitlab.

Yes @sbose78, Github is part of this PR, the generic git is available here https://github.com/redhat-developer/git-service/pull/7, the [GitLab is just waiting for tests](https://github.com/MatousJobanek/git-service/tree/gl-detector/pkg/git/repository/gitlab) and Bitbucket is waiting for the overall implementation

**Comment by kwk** [comment]
> As a 'first' pass, we should be good if we handled Github , Bitbucket and Gitlab.

In terms of effort and testing, why don't we support vanilla Git with it's protocols (https/ssh)? Then we can be sure that all work and we don't need any special handling for github, just because it has an API on which we might even hit a rate limit.

**Comment by alexeykazakov** [comment]
@kwk
The problem with vanilla git is that we have to clone when targeting vanilla git (we don’t have to clone when using GitHub/gitlab API).
And cloning unknown amount of data in the controller namespace is risky and not scalable. We would need to deploy a container in the user namespace for that so we are dealing with the usernames quota instead. So it’s much more complicated flow and we want to address it later.

**Comment by MatousJobanek** [comment]
@alexeykazakov ok, this is something new for me. I thought that the generic implementation should be present in the cases when we cannot use the API. To avoid running out of resources we wanted to use the user's namespace. At least this is what I understood from our last discussions.

So, even if we don't want to support vanilla git for now, I'd like to have the PR https://github.com/redhat-developer/git-service/pull/7 merged inside of the code anyway. We can disable it for now and then activate when we sort out the resources/namespaces problem. By merging it, I'd like to avoid situation with a lot of conflicts when we decide that we need it.

**Comment by alexeykazakov** [comment]
> To avoid running out of resources we wanted to use the user's namespace. At least this is what I understood from our last discussions.

This is exactly what I'm saying too, right? :)

> So, even if we don't want to support vanilla git for now, I'd like to have the PR #7 merged inside of the code anyway. We can disable it for now and then activate when we sort out the resources/namespaces problem. By merging it, I'd like to avoid situation with a lot of conflicts when we decide that we need it.

That's fine. Let's do it. We can have the detection service/code in place (PR #7). But we don't to want to use this code in the controller for now. The controller flow for vanila git will be more complicated (new deployment for the build type detector service in the user namespace, etc).

**Comment by sbose78** [comment]
Right, let's get the simple thing working first - which is support for
GitHub ( and Bitbucket, Gitlab ..) with their respective APIs.

The ideal situation is to always use something so generic that it works for
everything. That's what I wanted too, intially. Unfortunately , as I
understand, there's no easy way to accomplish it.

**Comment by xcoulon** [comment]
> > To avoid running out of resources we wanted to use the user's namespace. At least this is what I understood from our last discussions.
> 
> This is exactly what I'm saying too, right? :)
> 
> > So, even if we don't want to support vanilla git for now, I'd like to have the PR #7 merged inside of the code anyway. We can disable it for now and then activate when we sort out the resources/namespaces problem. By merging it, I'd like to avoid situation with a lot of conflicts when we decide that we need it.
> 
> That's fine. Let's do it. We can have the detection service/code in place (PR #7). But we don't to want to use this code in the controller for now. The controller flow for vanila git will be more complicated (new deployment for the build type detector service in the user namespace, etc).

If we want to run the `git clone && analysis` in the user's namespace, could we just spin a pod that would handle the job in the user's namespace ? In other words: the controller detects a new CR which contains the git repo metadata and starts a pod in the user namespace, so the operation will not run in the operator NS. Could it solve the problem of resource allocations while keeping the current design as-is ?

**Comment by MatousJobanek** [comment]
> This is exactly what I'm saying too, right? :)

@alexeykazakov yes, you are right of course :-). I just thought that we wanted to achieve it from the very first day, and not later. But I do understand that we should do the simple things first and the more complicated later, so I'm fine with it :+1:. There was just a small misunderstanding on my side :-)

> Could it solve the problem of resource allocations while keeping the current design as-is ?

exactly, you are right @xcoulon. However, the implementation will be probably a bit more complicated where we could face some additional complications that we don't realize right now. So, I guess that the @sbose78's point doing the generic detection later and for now just use the simple controller is totally valid.

**Comment by alexeykazakov** [comment]
The biggest question is how we would determine what namespace to use to run the generic git detection pod. Also it can be confusing for the user to see some weird pod in the namespace...
But let’s think about later when we have to implement it :)

**Pull request merged**